File: make_animation.py
#!/usr/bin/env python3
import math
import gzip
import constants
import subprocess
import os
from modules.bits import BitGrid

GRID_CACHE = {}
def get_grid(context, frame):
	#context = constants.TOPLEVEL_CONTEXT
	#context = 16 if context & 0x10 else 7
	#frame = 0
	if (context, frame) not in GRID_CACHE:
		with gzip.open(f"tiles/{context}/{frame}.gz", "rb") as fp:
			GRID_CACHE[context, frame] = BitGrid.load(fp)
	return GRID_CACHE[context, frame]

def clear_cache():
	GRID_CACHE.clear()

def split_range(valmin, valmax):
	ivalmin = math.floor(valmin)
	fvalmin = valmin - ivalmin
	ivalmax = math.ceil(valmax) - 1
	fvalmax = valmax - ivalmax
	if ivalmin == ivalmax:
		yield ivalmin, fvalmin * constants.LOOP_SCALE, fvalmax * constants.LOOP_SCALE
		return
	yield ivalmin, fvalmin * constants.LOOP_SCALE, constants.LOOP_SCALE
	for i in range(ivalmin + 1, ivalmax):
		yield i, 0.0, constants.LOOP_SCALE
	yield ivalmax, 0.0, fvalmax * constants.LOOP_SCALE

def is_covered(depth, frame, depthofs, depthmax, xmin, xmax, ymin, ymax, context=None):
	if depthofs >= 2:
		context = constants.TOPLEVEL_CONTEXT
		frameix = 0
	else:
		frameix = frame + constants.LOOP_LENGTH * (1 - depthofs)
	tile = get_grid(context, constants.FRAMES[frameix])
	if depthofs <= depthmax:
		return tile.is_covered((math.floor(xmin), math.floor(ymin)), (math.ceil(xmax), math.ceil(ymax)))
	else:
		splity = list(split_range(ymin, ymax))
		splitx = list(split_range(xmin, xmax))
		for y, newymin, newymax in splity:
			for x, newxmin, newxmax in splitx:
				if is_covered(depth, frame, depthofs - 1, depthmax,
						newxmin, newxmax, newymin, newymax, tile.context((x, y))):
					return True
		return False

def render_pixel_level(depth, frame, depthmax, xmin, xmax, ymin, ymax):
	for i in range(2):
		cx, cy = constants.CENTRES[depth + i]
		xmin = cx + (xmin / constants.LOOP_SCALE)
		xmax = cx + (xmax / constants.LOOP_SCALE)
		ymin = cy + (ymin / constants.LOOP_SCALE)
		ymax = cy + (ymax / constants.LOOP_SCALE)
	if is_covered(depth, frame, 2, depthmax, xmin, xmax, ymin, ymax):
		return 1.0
	else:
		return 0.0

def render_pixel(depth, frame, xmin, xmax, ymin, ymax):
	if depth >= 1 and frame < constants.INIT_NO_SCALE:
		p0 = render_pixel_level(depth, frame, 0, xmin, xmax, ymin, ymax)
		p1 = render_pixel_level(depth, frame, -1, xmin, xmax, ymin, ymax)
		prop = frame / constants.INIT_NO_SCALE
		return p1 * (1 - prop) + p0 * prop
	else:
		return render_pixel_level(depth, frame, 0, xmin, xmax, ymin, ymax)

def render_pixel_osa(depth, frame, xmin, xmax, ymin, ymax):
	dx = (xmax - xmin) / 2
	dy = (ymax - ymin) / 2
	cx = (xmin + xmax) / 2
	cy = (ymin + ymax) / 2
	bloomx = dx * constants.BLOOM
	bloomy = dx * constants.BLOOM
	if constants.OSA <= 1:
		return render_pixel(depth, frame, cx - bloomx, cx + bloomx, cy - bloomy, cy + bloomy)
	dat = 0
	for n in range(constants.OSA):
		px = xmin + dx * (2 * n + 1) / constants.OSA
		py = ymin + dy * (2 * n + 1) / constants.OSA
		dat += render_pixel(depth, frame, px - bloomx, px + bloomx, py - bloomy, py + bloomy)
		py = ymax - dy * (2 * n + 1) / constants.OSA
		dat += render_pixel(depth, frame, px - bloomx, px + bloomx, py - bloomy, py + bloomy)
	return dat / (2 * constants.OSA)

def render_frame(frame):
	depth, frame = divmod(frame, constants.LOOP_LENGTH)

	scale_frame = constants.INIT_NO_SCALE if depth <= 0 and frame <= constants.INIT_NO_SCALE else frame
	scalex = constants.SCALE_A * math.exp(constants.SCALE_K * scale_frame)
	scaley = scalex * constants.HEIGHT / constants.WIDTH

	centrex, centrey = constants.CENTRES_FLOAT[depth]
	xmin = centrex - scalex/2
	ymin = centrey - scaley/2

	for y in range(constants.HEIGHT):
		#print(f"{y=}")
		for x in range(constants.WIDTH):
			yield render_pixel_osa(depth, frame,
				xmin + scalex * (x / constants.WIDTH),
				xmin + scalex * ((x + 1) / constants.WIDTH),
				ymin + scaley * (y / constants.HEIGHT),
				ymin + scaley * ((y + 1) / constants.HEIGHT))

	clear_cache()

def save_frame(frame, dat):
	with open(f"frames/{frame:08d}.pgm", "w") as fp:
		print("P2", file=fp)
		print(f"{constants.WIDTH} {constants.HEIGHT} 255", file=fp)
		for ix, x in enumerate(dat):
			print(int(x * 255), file=fp, end=" ")
			if (ix + 1) % constants.WIDTH == 0:
				print(file=fp)
	subprocess.check_call(["convert", f"frames/{frame:08d}.pgm", f"frames/{frame:08d}.png"])
	os.unlink(f"frames/{frame:08d}.pgm")

def doframe(frame):
	print(frame)
	dat = render_frame(frame)
	save_frame(frame, dat)

def main():
	for i in range(0, 5160, 20):
		doframe(i)
	#doframe(0)

def main_profile():
	import cProfile
	import time
	prof = cProfile.Profile()
	prof.runcall(main)
	prof.dump_stats(time.strftime("%Y%m%d-%H%M%S") + ".profile")

if __name__ == '__main__':
	main()
	#main_profile()

File: make_animation.cpp
#include <iostream>
#include <fstream>
#include <iomanip>
#include <map>
#include <sstream>
#include <cmath>
#include <cstdlib>
#include <cstdio>
#include <ctime>
#include <unistd.h>
#include <sys/stat.h>

#include "bitgrid.h"
#include "constants.h"

std::map<std::pair<int, int>, BitGrid*> tilecache;
BitGrid &get_tile(int context, int frame) {
	std::pair<int,int> key = std::make_pair(context, frame);
	auto i = tilecache.find(key);
	if (i != tilecache.end())
		return *i->second;

	std::ostringstream filename;
	filename << "tiles/" << context << "/" << frame << ".gz";
	BitGrid *bg = BitGrid::load(filename.str());
	tilecache[key] = bg;
	return *bg;
}
void clear_tilecache() {
	for (auto i = tilecache.begin(); i != tilecache.end(); i++) {
		delete i->second;
	}
	tilecache.clear();
}

int ifloor(double);
int iceil(double);
int ifloor(double x) {
	if (x < 0)
		return -iceil(-x);
	return int(x);
}
int iceil(double x) {
	if (x < 0)
		return -ifloor(-x);
	int i = int(x);
	if (x > double(i))
		i++;
	return i;
}

bool is_covered(int depth, int frame, int depthofs, int depthmax, double x1, double x2, double y1, double y2, int context) {
	int frameix;
	if (depthofs >= 2) {
		frameix = 0;
	} else {
		frameix = frame + LOOP_LENGTH * (1 - depthofs);
	}
	BitGrid &tile = get_tile(context, FRAMES[frameix]);
	if (depthofs <= depthmax) {
		return tile.is_covered(ifloor(x1), iceil(x2), ifloor(y1), iceil(y2));
	} else {
		int ix1 = ifloor(x1), ix2 = ifloor(x2), iy1 = ifloor(y1), iy2 = ifloor(y2);
		double newx1 = (x1 - ix1) * LOOP_SCALE, newx2 = (x2 - ix2) * LOOP_SCALE, newy1 = (y1 - iy1) * LOOP_SCALE, newy2 = (y2 - iy2) * LOOP_SCALE;
		if (ix2 > ix1 + 1 || iy2 > iy1 + 1) {
			// shouldn't happen, sanity check shortcut
			return 1;
		}
		if (ix1 == ix2 && iy1 == iy2) {
			return is_covered(depth, frame, depthofs - 1, depthmax, newx1, newx2, newy1, newy2, tile.context(ix1, iy1));
		} else if (ix1 == ix2) {
			return (
				is_covered(depth, frame, depthofs - 1, depthmax, newx1, newx2, newy1, LOOP_SCALE, tile.context(ix1, iy1)) ||
				is_covered(depth, frame, depthofs - 1, depthmax, newx1, newx2, 0, newy2, tile.context(ix1, iy2))
			);
		} else if (iy1 == iy2) {
			return (
				is_covered(depth, frame, depthofs - 1, depthmax, newx1, LOOP_SCALE, newy1, newy2, tile.context(ix1, iy1)) ||
				is_covered(depth, frame, depthofs - 1, depthmax, 0, newx2, newy1, newy2, tile.context(ix2, iy1))
			);
		} else {
			return (
				is_covered(depth, frame, depthofs - 1, depthmax, newx1, LOOP_SCALE, newy1, LOOP_SCALE, tile.context(ix1, iy1)) ||
				is_covered(depth, frame, depthofs - 1, depthmax, 0, newx2, newy1, LOOP_SCALE, tile.context(ix2, iy1)) ||
				is_covered(depth, frame, depthofs - 1, depthmax, newx1, LOOP_SCALE, 0, newy2, tile.context(ix1, iy2)) ||
				is_covered(depth, frame, depthofs - 1, depthmax, 0, newx2, 0, newy2, tile.context(ix2, iy2))
			);
		}
	}
}

unsigned char render_pixel_level(int depth, int frame, int depthmax, double x1, double x2, double y1, double y2) {
	x1 = CENTRESX[depth] + x1 / LOOP_SCALE;
	x2 = CENTRESX[depth] + x2 / LOOP_SCALE;
	y1 = CENTRESY[depth] + y1 / LOOP_SCALE;
	y2 = CENTRESY[depth] + y2 / LOOP_SCALE;

	x1 = CENTRESX[depth+1] + x1 / LOOP_SCALE;
	x2 = CENTRESX[depth+1] + x2 / LOOP_SCALE;
	y1 = CENTRESY[depth+1] + y1 / LOOP_SCALE;
	y2 = CENTRESY[depth+1] + y2 / LOOP_SCALE;

	return is_covered(depth, frame, 2, depthmax, x1, x2, y1, y2, TOPLEVEL_CONTEXT) ? 255 : 0;
}

unsigned char render_pixel(int depth, int frame, double x1, double x2, double y1, double y2) {
	if (depth >= 1 && frame < INIT_NO_SCALE) {
		int p0 = render_pixel_level(depth, frame, 0, x1, x2, y1, y2);
		int p1 = render_pixel_level(depth, frame, -1, x1, x2, y1, y2);
		return (p1 * (INIT_NO_SCALE - frame) + p0 * frame) / INIT_NO_SCALE;
	} else {
		return render_pixel_level(depth, frame, 0, x1, x2, y1, y2);
	}
}

unsigned char render_pixel_osa(int depth, int frame, double x1, double x2, double y1, double y2) {
	double dx = (x2 - x1) / 2;
	double dy = (y2 - y1) / 2;
	double cx = (x1 + x2) / 2;
	double cy = (y1 + y2) / 2;
	double bloomx = dx * BLOOM;
	double bloomy = dy * BLOOM;
	if (OSA > 1) {
		// do over-sampling
		int dat = 0;
		for (int i = 0; i < OSA; i++) {
			double factor = 1.0 + double(i) / OSA;
			dat += render_pixel(depth, frame, cx - bloomx * factor, cx + bloomx * factor, cy - bloomy * factor, cy + bloomy * factor);
		}
		return dat / OSA;
	} else {
		// single-sample
		return render_pixel(depth, frame, cx - bloomx, cx + bloomx, cy - bloomy, cy + bloomy);
	}
}

unsigned char *render_frame(int frame) {
	int depth = frame / LOOP_LENGTH;
	frame = frame % LOOP_LENGTH;

	int scale_frame = depth <= 0 && frame < INIT_NO_SCALE ? INIT_NO_SCALE : frame;
	double scalex = SCALE_A * std::exp(SCALE_K * scale_frame);
	double scaley = scalex * HEIGHT / WIDTH;

	double xmin = CENTRES_FLOATX[depth] - scalex/2;
	double ymin = CENTRES_FLOATY[depth] - scaley/2;

	static unsigned char image_buffer[WIDTH * HEIGHT];
	unsigned char *ptr = image_buffer;

	for (int y = 0; y < HEIGHT; y++)
		for (int x = 0; x < WIDTH; x++)
			*ptr++ = render_pixel_osa(depth, frame, xmin + scalex * double(x)/WIDTH, xmin + scalex * double(x+1)/WIDTH, ymin + scaley * double(y)/HEIGHT, ymin + scaley * double(y+1)/HEIGHT);

	clear_tilecache();

	return image_buffer;
}

void save_frame(int frame, unsigned char *image) {
	std::ostringstream filename;
	filename << "frames/" << std::setw(4) << std::setfill('0') << (frame / LOOP_LENGTH);
	if (access(filename.str().c_str(), F_OK)) {
		if (mkdir(filename.str().c_str(), 0755)) {
			std::cerr << "Error creating subdir" << std::endl;
			std::exit(1);
		}
	}
	filename << '/' << std::setw(8) << std::setfill('0') << frame;
	std::string pgmfilename = filename.str() + ".pgm";
	std::string pngfilename = filename.str() + ".png";

	std::ofstream fp;
	fp.exceptions(std::ofstream::failbit | std::ofstream::badbit);
	fp.open(pgmfilename);

	fp << "P2\n" << WIDTH << " " << HEIGHT << " 255\n";
	unsigned char *ptr = image;
	for (int y = 0; y < HEIGHT; y++) {
		for (int x = 0; x < WIDTH; x++) {
			fp << int(*ptr++) << " ";
		}
		fp << "\n";
	}
	fp.close();

	std::ostringstream cmdline;
	cmdline << "convert " << pgmfilename << " " << pngfilename;
	if (std::system(cmdline.str().c_str())) {
		std::cerr << "Error running convert" << std::endl;
		std::exit(1);
	}
	if (std::remove(pgmfilename.c_str())) {
		std::cerr << "Error unlinking pgm file" << std::endl;
		std::exit(1);
	}
}

void doframe(int frame) {
	std::time_t now = time(NULL);
	std::string nowstr = std::ctime(&now);
	size_t pos = nowstr.find_last_not_of('\n');
	if (pos == std::string::npos)
		nowstr = "";
	else if (pos < nowstr.length() - 1)
		nowstr.erase(pos + 1, std::string::npos);

	std::cout << frame << " - " << nowstr << std::endl;

	unsigned char *image = render_frame(frame);
	save_frame(frame, image);
}

void doanim(int ofs, int step) {
	for (int i = ofs; i < LOOP_LENGTH * LOOP_COUNT; i += step) {
		doframe(i);
	}
}

void usage() {
	std::cout
		<< "Usage:\n"
		<< "  make_animation              - render entire animation\n"
		<< "  make_animation frame        - render a single frame\n"
		<< "  make_animation offset step  - render a slice of the animation\n";
}

int main(int argn, char **argv) {
	if (argn == 1) {
		doanim(1, 1);
	} else if (argn == 2) {
		char *endptr;
		int frame = std::strtol(argv[1], &endptr, 0);
		if (*endptr != '\0') {
			usage();
			return 1;
		}

		doframe(frame);
	} else if (argn == 3) {
		char *endptr;
		int ofs = std::strtol(argv[1], &endptr, 0);
		if (*endptr != '\0') {
			usage();
			return 1;
		}
		int step = std::strtol(argv[2], &endptr, 0);
		if (*endptr != '\0') {
			usage();
			return 1;
		}

		doanim(ofs, step);
	} else {
		usage();
	}

	return 0;
}
